src/vs/workbench/contrib/chat/browser/chatVariables.ts

```diff
@@ -79,30 +149,69 @@ export class ChatVariablesService implements IChatVariablesService {
 						}
 						progress(item);
 					};
-					jobs.push(data.resolver(prompt.text, '', model, variableProgressCallback, token).then(value => {
-						if (value) {
-							resolvedAttachedContext[i] = { id: data.data.id, modelDescription: data.data.modelDescription, name: attachment.name, fullName: attachment.fullName, range: attachment.range, value, references, icon: attachment.icon };
+
+					try {
+						const value = await data.resolver(prompt.text, '', model, variableProgressCallback, token);
+						if (!value) {
+							return null;
 						}
-					}).catch(onUnexpectedExternalError));
-				} else if (attachment.isDynamic || attachment.isTool) {
-					resolvedAttachedContext[i] = attachment;
+
+						return {
+							id: data.data.id,
+							modelDescription: data.data.modelDescription,
+							name: attachment.name,
+							fullName: attachment.fullName,
+							range: attachment.range,
+							value,
+							references,
+							icon: attachment.icon,
+						};
+					} catch (error) {
+						onUnexpectedExternalError(error);
+
+						throw error;
+					}
+				}
+
+				if (attachment.isDynamic || attachment.isTool) {
+					return attachment;
 				}
-			});
 
-		await Promise.allSettled(jobs);
+				return null;
+			});
 
-		// Make array not sparse
-		resolvedVariables = coalesce<IChatRequestVariableEntry>(resolvedVariables);
+		// run all jobs in paralle and get results in the original order
+		// Note! the `getJobResults` call supposed to never fail, hence it's ok to do
+		// 	     `Promise.all()`, otherwise we have a logic error that would be caught here
+		const [resolvedVariables, resolvedAttachedContext] = await Promise.all(
+			[
+				getJobResults(resolvedVariableJobs),
+				getJobResults(resolvedAttachedContextJobs),
+			],
+		);
 
-		// "reverse", high index first so that replacement is simple
-		resolvedVariables.sort((a, b) => b.range!.start - a.range!.start);
+		// "reverse" resolved variables making the high index
+		// to go first so that an replacement logic is simple
+		resolvedVariables
+			.sort((left, right) => {
+				assertDefined(
```

Comment:
These asserts are to prevent TS hacks ☺️

Reply:
If they always have ranges, we need to update the TS types then?
The asserts just add a more readable error messages here as opposed to a cryptic `cannot read 'start' of 'undefined'` you'd get with the bang ☺️

Reply:
That type doesn't always have ranges, but here you're just working with prompt variables